### PR TITLE
ppx

### PR DESCRIPTION
**Pull request by smolkaj**
I'm creating this pull request already so we can keep track of what needs to be done, @jnfoster:

- [x] Parser
  - [x] integrate with oasis
  - [x] add support for local fields
  - [x] use proper source locations when generating OCaml AST (see [metaloc](https://github.com/alainfrisch/ppx_tools/blob/master/ppx_metaquot.ml))
  - [x] switch to unsigned int32 and int64
     -> create separate issue

- [x] Lexer
  - [x] Lexical convention for meta ids to avoid hard-to-diagnose typo bugs?
     -> all meta field names must start with '`'

- [x] ppx
  - [x] support let-expressions: `let%nk p = {| ... |} in e`
  - [x] support declarations: `let%nk p = {| ... |}`
  - [x] support constants: `{nk| port:=1 |nk}` ?
    -> probably not a great idea, because such a constant will be parsed as a string without warnings if syntax extension is accidentally not enabled
- [x] update opam and oasis files

- [x] insert proper source locations when building NetKAT AST
- [x] open Frenetic_NetKAT automatically
- [x] catch errors (dump command, async)
- [x] update examples: new syntax for meta fields and union

### CODE REVIEW

**Comment by smolkaj** [comment]
This is almost ready to be merged.

**Comment by smolkaj** [comment]
This is ready to be merged.

@basus and @jnfoster, can you give it a try and give some feedback? There is a test file `main.ml` in `ppx/test`. To execute it, simply run `make` in that directory. (This will first build and install frenetic. The syntax extension can only be used once it is installed as a findlib library.)

Alternatively, if you want to see the source file after rewriting, go to the `_opam` file, locate `ppx_standalone`, and set `Build: true`. This will build an executable that takes in a `*.ml` file that may contain NetKAT syntax, and outputs an `*.ml` file where the NetKAT syntax has been replaced with the appropriate OCaml code. 

Some things to note:
* I decided that meta fields should always start with a back tick. This is to avoid horrible bugs where you write `ip4src` instead of `ip4Src`, and the compiler (without you noticing) interprets the field as a local field.
* I also added a `let%nk_pred ...` version for predicates.
* lex and parse errors should produce error messages with precise source locations
* I decided to use `+` instead of `|`. It's trivial to change that, if we want.

**Comment by smolkaj** [comment]
@jnfoster, @basus: There is some annoying technical issue with menhir that keeps travis from building the branch successful. I haven't been able to figure out what the problem is or how to fix it. Any ideas?

**Comment by basus** [comment]
I don't have any experience with menhir or travis, I'm afraid. I like the idea of syntactically separating meta and standard fields.

**Comment by smolkaj** [comment]
@basus: Did you try building it? Did it work on your machine?

**Comment by basus** [comment]
Pushed a change that should make things compile. The name of the Menhir library for `findlib` is `menhirLib`, not `menhirlib`.

**Comment by basus** [comment]
This branch builds fine on both my OS X Yosemite machine, and on my Arch Linux machine. Looking at the Travis output, the documentation building now fails with the following output:

```
ocaml setup.ml -doc
Finished, 1 target (0 cached) in 00:00:00.
Solver failed:
  Ocamlbuild knows of no rules that apply to a target named lib/frenetic.odocl. This can happen if you ask Ocamlbuild to build a target with the wrong extension (e.g. .opt instead of .native) or if the source files live in directories that have not been specified as include directories.
Compilation unsuccessful after building 0 targets (0 cached) in 00:00:00.
E: Failure("Command ''/home/travis/.opam/4.03.0/bin/ocamlbuild' lib/frenetic.docdir/index.html -use-ocamlfind -tag debug' terminated with error code 6")
make: *** [doc] Error 1
```

**Comment by smolkaj** [comment]
Awesome @basus, you did it! 👍 👍 👍 
It worked on my machine, probably because MacOS is case-insensitive (but Linux is not).

The new problem is probably easy to fix.

**Comment by baxtersa** [comment]
I just fixed some syntax bugs in the ppx test file from when you changed the lexer. Otherwise, this looks good on my machine.

@smolkaj What's the status of remaining work on this branch? Once it lands I can take a look at updating the netkat tutorial examples if that would be helpful.

**Comment by smolkaj** [comment]
Thanks, @baxtersa. The main holdback is that Travis is not happy with the current setup, because the files `lib/Frenetic_NetKAT_Tokens.{ml,mli}` are auto-generated and so ocamldoc cannot find them. Not sure what's the best way to work around this - maybe simple check in those files?

**Comment by baxtersa** [comment]
There's a `PreCommand` field for `Document` sections in an `oasis` file that we can use to run `cppo` and `menhir` before running ocamldoc. It fixes the issue for me locally, so we'll see if it makes travis happy.

**Comment by smolkaj** [comment]
@baxtersa: sweet, thanks for figuring this out!

File: lib/Parser.cppo.mly [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+  | LPAR; a=pred; RPAR
+      { a }
+  | BEGIN a=pred; END
+      { a }
+  AQ
+  ;
+
+
+(*********************** HEADER VALUES *************************)
+(* FIXME: add local fields (aka meta fields) back to the language! *)
```
Comment (smolkaj) [inline lib/Parser.cppo.mly]: remove this

File: lib/Parser.cppo.mly [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    BOTH( IPProto (Obj.magic 0) )
+  | IP4SRC
+    BOTH( IP4Src (Obj.magic 0, Obj.magic 0) )
+  | IP4DST
+    BOTH( IP4Dst (Obj.magic 0, Obj.magic 0) )
+
+
+(*********************** VALUES *************************)
+(* FIXME: we should be parsing *unsigned* int32 and int64 here! *)
+
```
Comment (smolkaj) [inline lib/Parser.cppo.mly]: open separate issue

File: lib/Parser.cppo.mly [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+%{
+open Core.Std
+open Frenetic_NetKAT
+#ifdef MAKE_PPX
+open Parsetree
+open Ast_helper
+open Ast_convenience
+(* FIXME: temporary, until these are included in the next Ast_convenience release *)
+let int32 ?loc ?attrs x =
```
Comment (smolkaj) [inline lib/Parser.cppo.mly]: open separate issue

File: _oasis [hunk 4]
```diff
@@ -5,57 +5,76 @@ Version:     4.0.0
...
-    syntax/Frenetic_Syntax_Extension_Parser.cppo.ml -o \
-    syntax/Frenetic_Syntax_Extension_Parser.ml
+  cppo -n \
+    lib/Frenetic_NetKAT_Generated_Parser.cppo.mly \
+    -o lib/Frenetic_NetKAT_Generated_Parser.mly
+  cppo -n \
+    ppx/PPX_NetKAT_Parser.cppo.mly \
+    -o ppx/PPX_NetKAT_Parser.mly
+  menhir --only-tokens ppx/PPX_NetKAT_Parser.mly \
+    --base lib/Frenetic_NetKAT_Tokens
```
Comment (smolkaj) [inline _oasis]: can we move this to the place where it is actually needed?

File: lib/Frenetic_NetKAT_Lexer.ml [hunk 5]
```diff
@@ -1,186 +1,169 @@
...
+  | "or" -> OR
+  | "not" -> NOT
+  | '=' -> EQUALS
+  (* policies *)
+  | "id" -> ID
+  | "drop" -> DROP
+  | "filter" -> FILTER
+  | ":=" -> ASSIGN
+  | ';' -> SEMICOLON
+  (* SJS: let's start using `+` instead of `|` ? *)
```
Comment (smolkaj) [inline lib/Frenetic_NetKAT_Lexer.ml]: Use `+` instead of `|`, as discussed with @jnfoster

File: ppx/test/main.ml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+(* can have open terms *)
+let%nk r = {| `inport := 1 |}
+
+(*  *)
+let%nk r = {| `aux := 192.168.2.1; filter `aux = 00:0a:95:9d:68:16 |}
+
+(* ERRORS *)
+(* let%nk s = {| filter typo = 1 |} *)
+(* let%nk r = {| while not $egress' do $q |} *)
+(* let%nk r = {| `inport := port |} *)
```
Comment (smolkaj) [inline ppx/test/main.ml]: Make sure these tests cover all four extensions: `let x = e` and `let x = e1 in e2` with `%nk` and `%nk_pred`

File: _oasis (hunk 4 above)
Comment (smolkaj) [inline _oasis]: nope, oasis does not support `{Pre,Post}BuildCommand` for `library`s

**Comment by baxtersa** [comment]
The parser doesn't handle `vswitch`/`vport` tokens. I'm working on add necessary virtual compiler support and updating all the examples.

**Comment by baxtersa** [comment]
Looks like a new mparser release broke the build.

**Comment by baxtersa** [comment]
Looks like a new mparser release broke travis.

**Comment by basus** [comment]
I filed a bug for the MParser issue on the OPAM issue tracker a few days ago, but it hasn't gotten any attention. Maybe you guys can help bump it up?
https://github.com/ocaml/opam-repository/issues/8000

**Comment by baxtersa** [comment]
See https://github.com/cakeplus/mparser/issues/6 for discussion of the issue.

Long story short the `mparser` author is issuing a fix for it, so I'll revert my version pinning (which didn't work with travis anyway).

**Comment by baxtersa** [comment]
Things are looking good again!

**Comment by baxtersa** [comment]
Any feedback on this PR now? I'm happy to continue checking off issues to get this merged.

**Comment by smolkaj** [comment]
@baxtersa: Thanks a lot for getting this done! Amazing 👍 
I'm a little busy right now, so let's just merge this in but let the PR open. I'll take a final look once I have the time.

**Comment by baxtersa** [comment]
Great! Thanks for doing the heavy lifting.

**Pull request merged**